### DISCUSSION TITLE
Progress chart

### DISCUSSION THREAD

**Discussion by mattiason** [Budibase/budibase#5997] [Help]
Hi

I am trying to create a progress chart with help of donut chart. 

I have a table with one column for completed items and one for total items. 

I would like to show the progress. Is this possible?
<img width="207" alt="Screenshot 2022-05-20 at 09 34 55" src="https://user-images.githubusercontent.com/45860668/169477572-bdcdb598-090c-4394-90bf-cc33eec56168.png">
<img width="58" alt="Screenshot 2022-05-20 at 09 34 51" src="https://user-images.githubusercontent.com/45860668/169477577-6cd9b953-8fa4-437e-987c-332682674414.png">

**Comment by melohagan** [comment]
Hi @mattiason 

Yes this should possible; you will want your schema to be in this format: https://docs.budibase.com/docs/chart#generating-the-chart-schema

So you'll need a query that returns something approximating:
```
[ { label: "Completed", value: data.completedItems }, { label: "Not Completed", value: data.nrOfItems - data.completedItems } ]
```

Then if you wish, you can hide the data labels and legend, by unchecking these:
![Screenshot 2022-05-20 at 10 04 16](https://user-images.githubusercontent.com/101575380/169494431-08fdff35-0ef4-49bc-b574-ce16face81ba.png)

Have a play around with the colour palette too.

**Reply by mattiason** [reply]
Okey, so it is only possible through query? At the moment for business reasons, we are updating one record so it is only 1 record with 2  number fields.

So I need to have as sub queries?

**Reply by melohagan** [reply]
You should be able to use a custom query to get the data into the correct format: https://docs.budibase.com/docs/custom-queries

If you're using the internal table, then you can use a view to group your tasks by the count of their completion status:
![Screenshot 2022-05-20 at 10 32 44](https://user-images.githubusercontent.com/101575380/169499631-ca5b5e53-f57c-4538-a3df-2ad35a1f646d.png)

![Screenshot 2022-05-20 at 10 31 10](https://user-images.githubusercontent.com/101575380/169499381-bac4a191-d5ab-4d4a-83e8-6805d011a688.png)

![Screenshot 2022-05-20 at 10 34 18](https://user-images.githubusercontent.com/101575380/169499913-0714b672-4bc6-412d-951d-7d3a3575d4b5.png)

Where the New Data Provider is using the view as its data source.

**Reply by mattiason** [reply]
Okey I see. This gets the issue is that we don't do it over multiple records we only have 2 fields 1 said no items and 1 completed items and we would like to use one as max value and the other to show progress. We have also tried to add a calculation field showing progress. There are operational reasons why we don't want to add new record for everything. 

<img width="506" alt="image" src="https://user-images.githubusercontent.com/45860668/169500835-db630d58-ce60-47f2-8e12-85f629b7ba8e.png">

**Reply by melohagan** [reply]
I see, is this using the internal Budibase table? 
If it was an external table, you could use a custom query to transform the data above into the correct format.

**Reply by mattiason** [reply]
Yes it is using the internal table in this case. 
So you mean I could achieve it using e.g SQL db but with the same principle storing in the event?

**Reply by melohagan** [reply]
Yeah so if it was a SQL table then you could have a query and transformer roughly:

**Query**
`SELECT * FROM Table`

**Transformer**
```
return [ { label: "Completed", value: data.completedItems }, { label: "Not Completed", value: data.nrOfItems - data.completedItems } ]
```
*as mentioned above*

**Reply by mattiason** [reply]
but what if i use a json field in bud base internal table like progressdata: [ { label: "Completed", value: data.completedItems }, { label: "Not Completed", value: data.nrOfItems - data.completedItems } ]. Could that work? Since i understand that arrays in json could be used as data sources? Would this be possible?

**Reply by melohagan** [reply]
It could, although you'd need to store the actual values e.g.: ` [ { label: "Completed", value: 2 }, { label: "Not Completed", value: 11 } ]`

**Reply by mattiason** [reply]
Thanks will give it a try later
